torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +111,184 @@ def load_actor_json(filename: str) -> List[RayActor]:
         return actors
 
 
-def create_placement_group(replicas: List[RayActor]) -> PlacementGroup:
+def create_placement_group_async(replicas: List[RayActor]) -> PlacementGroup:
+    """return a placement group reference, the corresponding placement group could be scheduled or pending"""
     bundles = []
     for replica in replicas:
         bundles.append({"CPU": replica.num_cpus, "GPU": replica.num_gpus})
 
-    # To change the strategy type
-    # refer to available options here https://docs.ray.io/en/latest/placement-group.html#pgroup-strategy
     pg = ray.util.placement_group(bundles, strategy="SPREAD")
+    return pg
 
-    _logger.info("Waiting for placement group to start.")
-    ready = pg.wait(timeout_seconds=100)
 
-    if not ready:  # pragma: no cover
-        raise TimeoutError(
-            "Placement group creation timed out. Make sure "
-            "your cluster either has enough resources or use "
-            "an autoscaling cluster. Current resources "
-            "available: {}, resources requested by the "
-            "placement group: {}".format(ray.available_resources(), pg.bundle_specs)
+@dataclass
+class ActorInfo:
+    pg: PlacementGroup
+    replica: RayActor
+    actor: CommandActor
+
+
+class RayDriver:
+    def __init__(self, replicas: List[RayActor]) -> None:
+        self.replicas = replicas
+        self.master_node_id: Optional[str] = None
+        self.rank_0_address: Optional[str] = None
+        self.rank_0_port: Optional[int] = None
+        self.min_nnodes, self.max_nnodes = parse_nnodes_rep(replicas)  # pyre-ignore
+
+        self.placement_groups: List[PlacementGroup] = []
+        self.actor_info_of_id: Dict[str, ActorInfo] = {}
+        self.active_tasks: List["ray.ObjectRef"] = []
+
+    def init_placement_groups(self) -> None:
+        """Initialize all placement groups needed for this job"""
+        replica_ix_of_pg: List[int] = [0] + list(
+            range(self.min_nnodes, self.max_nnodes + 1)
         )
-    return pg
+        # trace all the placement groups, {placemeng_group_reference: placement_group_index}
+        self.placement_groups = [
+            create_placement_group_async(
+                self.replicas[replica_ix_of_pg[i] : replica_ix_of_pg[i + 1]]
+            )
+            for i in range(len(replica_ix_of_pg) - 1)
+        ]
+
+    def pop_actor_info(self, actor_id: str) -> ActorInfo:
+        """Remove the info of  (dead) command actor"""
+        return self.actor_info_of_id.pop(actor_id)
 
+    def reschedule_actor(self, actor_id: str) -> None:
+        """Rescheule a failed actor"""
```

Comment:
we should check the role max_retries/retry_policy here and throw an error if any of the workers exits more than N times

Reply:
Actually, here if we add a max_retries, it actually means how many node failures we can endure, so it's different from the max_retries in the ray actor's context.

Reply:
As long as the number of working nodes is bigger than min_nnodes, we should allow infinite node failures, unless nodes are failing too frequently. That can bea retry_policy.